File: LeetCode_Algorithm/215(Kth Largest Element in an Array)/215.py
import heapq
from typing import List
class Heap():
    def __init__(self, items):
        self.items = items
        # 중간에서 부터 뒤에 있는 자식 노드를 확인한다.
        for idx in range(len(self.items)//2, -1, -1):
            self.makeheap(self.items, idx)

    def makeheap(self, items, idx):
        largest = idx
        left = 2*largest+1
        right = 2*largest+2
        # 가장 큰 값이 오도록 바꿈
        if left < len(self.items) and items[left] > items[largest]:
            largest = left
        if right < len(self.items) and items[right] > items[largest]:
            largest = right
        if largest != idx:
            items[largest], items[idx] = items[idx], items[largest]
            self.makeheap(items, largest)

    def heap_add(self, num):
        self.items.append(num)
        for idx in range(len(self.items)//2, -1, -1):
            self.makeheap(self.items, idx)
    def heap_pop(self):
        num = self.items.pop(0)
        for idx in range(len(self.items)//2, -1, -1):
            self.makeheap(self.items, idx)
        return num

class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        # # sol1 time limit exceeded 5000개
        # # O(n)으로 k번째로 큰 수 찾기
        # heap = Heap(nums)
        # # max heap을 구현해서 그만큼 pop
        # for _ in range(k):
        #     num = heap.heap_pop()
        # return num
        
        # sol2 random select
        # O(n) 875ms 25%
        # if not nums: return ;
        # import random
        # pivot = random.choice(nums)
        # left = [x for x in nums if x > pivot]
        # mid = [x for x in nums if x == pivot]
        # right = [x for x in nums if x < pivot]
        # l = len(left); m = len(mid)
        # # k가 임의의 개수보다 더 작으면
        # if k <= l:
        #     return self.findKthLargest(left, k)
        # # 현재 k가 더 크다면 그만큼 큰 수 만큼 제거 한뒤 
        # elif k > l+m:
        #     return self.findKthLargest(right, k-l-m)
        # else:
        #     return mid[0]

        # sol3 heapq
        # 459ms 86% 29.27MB 89%
        import heapq
        heap = []
        for num in nums:
            heapq.heappush(heap, num)
            if len(heap) > k:
                heapq.heappop(heap)
        return heap[0]

print(Solution().findKthLargest([3,2,1,5,6,4], 2))

File: LeetCode_Algorithm/215(Kth Largest Element in an Array)/215.js
class Heap {
  constructor() {
    this.heap = []
  }
  parent(idx) {
    return Math.floor((idx - 1) / 2)
  }
  leftChild(idx) {
    return (2 * idx + 1)
  }
  rightChild(idx) {
    return (2 * idx + 2)
  }
  hasleftChild(idx) {
    return (2 * idx + 1) < this.heap.length
  }
  hasrightChild(idx) {
    return (2 * idx + 2) < this.heap.length
  }
  heappush(num) {
    this.heap.push(num)
    let idx = this.heap.length-1
    let par = this.parent(idx)
    while (par >= 0 && this.heap[par] > this.heap[idx]) {
      this.swap(par, idx)
      idx = par
      par = this.parent(idx)
    }
  }
  heappop() {
    const ret= this.heap[0]
    this.heap[0] =  this.heap[this.heap.length-1]
    this.heap.pop()
    let idx = 0
    while (this.hasleftChild(idx)) {
      let small = this.leftChild(idx)
      if (this.hasrightChild(idx) && this.heap[small] >= this.heap[this.rightChild(idx)]) {
        small = this.rightChild(idx)
      }
      if (this.heap[idx] >= this.heap[small]) {
        this.swap(idx, small)
      } 
      idx = small
    }
    return ret
  }
  swap(par, idx) {
    let temp = this.heap[par]
    this.heap[par] = this.heap[idx]
    this.heap[idx] = temp
  }
}

var findKthLargest = function(nums, k) {
  // sol2 runtime error 5000개중 첫번째로 큰 수 찾기
  // let pivot = nums[Math.floor(Math.random()) * nums.length()]
  // let left = []; let mid = []; let right = []
  // for (let num of nums) {
  //   if (num < pivot) {
  //     left.push(num)
  //   } else if (num == pivot) {
  //     mid.push(num)
  //   } else {
  //     right.push(num)
  //   }
  // }
  // if (k < left.length()) {
  //   return findKthLargest(left, k)
  // } else if (k > mid.length()+right.length()) {
  //   return findKthLargest(right, k-mid.length()-right.length())
  // } else {
  //   return mid[0]
  // }

  // sol1 time limit exceeded 63565번째 큰 수 찾기
  //  let heap = new Heap(nums)
  //   for (let idx=1; idx < k; idx++) {
  //     heap.items.pop()
  //   } return heap.items[0]

  // 98ms 87% 50.8MB 71%
  let heap = new Heap()
  for (let num of nums) {
      heap.heappush(num)
      if (heap.heap.length > k) 
        heap.heappop()
  }
  return heap.heap[0]
};

// console.log(findKthLargest([4,2,3,5,1], 2))
console.log(findKthLargest([7,6,5,4,3,2,1]  ,5))
